File: dynasty-scout/lib/dynastyValues.ts
/**
 * Dynasty trade values, refreshed where the app actually runs.
 *
 * The server-side twin of scrapers/rankings/dynasty_values.py. The Python one
 * writes to the local SQLite file, which is the right thing on this machine
 * and reaches nothing in production — Vercel runs against Supabase and cannot
 * run Python. So the daily refresh lives here, behind a cron route, and the
 * script stays for working locally.
 *
 * Two things this deliberately does not do.
 *
 * It does not touch the rookie sources. "KeepTradeCut", "FantasyPros",
 * "DynastyNerds" and "FantasyCalc" hold how the 2026 class was seen before
 * the draft; run_consensus.py weights them into the rookie board, and they
 * are a snapshot worth keeping rather than a feed worth refreshing. Dynasty
 * values are their own sources.
 *
 * And it never deletes. Every write is keyed on (player_id, source,
 * scraped_at), so a second run on one day corrects that day and every earlier
 * day survives. That is what makes a player's value a line rather than a
 * number: you can only see that somebody has been sliding for a month if the
 * month is still there.
 */
import { query, getDb } from '@/lib/db';

const BASE = 'https://api.fantasycalc.com/values/current';
const SOURCE_URL = 'https://fantasycalc.com/rankings';

/** The two formats, and which column of `dynasty_picks` each one fills. */
export const DYNASTY_FEEDS = [
    {
        key: '1qb',
        source: 'FantasyCalc Dynasty',
        query: '?isDynasty=true&numQbs=1&ppr=1&superflex=false',
        valueCol: 'value_1qb',
        rankCol: 'rank_1qb',
    },
    {
        key: 'sf',
        source: 'FantasyCalc Dynasty SF',
        query: '?isDynasty=true&numQbs=2&ppr=1&superflex=true',
        valueCol: 'value_sf',
        rankCol: 'rank_sf',
    },
] as const;

/** "2027 1st (Early)", "2028 2nd", … */
const PICK_LABEL = /^(\d{4}) (\d+)(?:st|nd|rd|th)(?: \((Early|Mid|Late)\))?$/;

interface FcRow {
    player: {
        name: string;
        position: string;
        sleeperId?: string | null;
    };
    value: number;
    overallRank: number | null;
    positionRank: number | null;
}

/**
 * Picks are assets, not players, so they get their own table rather than a
 * fake row in `players` that every other query would then have to exclude.
 */
export async function ensurePickTable(): Promise<void> {
    // run(), not query(): `query` reads with .all(), which SQLite refuses for
    // a statement that returns no rows, so the local path 500s on the DDL.
    await getDb().prepare(
        `CREATE TABLE IF NOT EXISTS dynasty_picks (
           label      TEXT PRIMARY KEY,
           season     INTEGER NOT NULL,
           round      INTEGER NOT NULL,
           slot       TEXT,
           value_1qb  INTEGER,
           value_sf   INTEGER,
           rank_1qb   INTEGER,
           rank_sf    INTEGER,
           scraped_at TEXT NOT NULL
         )`).run([]);
}

async function fetchFeed(q: string): Promise<FcRow[]> {
    const res = await fetch(BASE + q, {
        headers: { 'user-agent': 'dynasty-scout/1.0' },
        cache: 'no-store',
    });
    if (!res.ok) throw new Error(`${res.status} from api.fantasycalc.com`);
    return res.json() as Promise<FcRow[]>;
}

async function savePlayers(
    rows: FcRow[], source: string, bySleeper: Map<string, number>, today: string,
): Promise<number> {
    const matched = rows
        .filter(r => r.player.position !== 'PICK')
        .map(r => ({ row: r, pid: bySleeper.get(String(r.player.sleeperId ?? '')) }))
        .filter((x): x is { row: FcRow; pid: number } => x.pid != null)
        // Two feed rows resolving to one player would break the 1..N
        // ordering below, so the better rank wins.
        .sort((a, b) => (a.row.overallRank ?? 9e9) - (b.row.overallRank ?? 9e9));

    const seen = new Set<number>();
    const dense = matched.filter(x => !seen.has(x.pid) && seen.add(x.pid));

    for (let i = 0; i < dense.length; i += 200) {
        const chunk = dense.slice(i, i + 200);
        const params: (number | string | null)[] = [];
        const values = chunk.map((x, j) => {
            const b = params.length;
            params.push(x.pid, source, i + j + 1, x.row.positionRank,
                SOURCE_URL, today, x.row.value);
            return `($${b + 1},$${b + 2},$${b + 3},$${b + 4},$${b + 5},$${b + 6},$${b + 7})`;
        });
        await getDb().prepare(
            `INSERT INTO rankings
               (player_id, source, rank_overall, rank_positional,
                source_url, scraped_at, value)
             VALUES ${values.join(',')}
             ON CONFLICT (player_id, source, scraped_at) DO UPDATE SET
               rank_overall    = excluded.rank_overall,
               rank_positional = excluded.rank_positional,
               source_url      = excluded.source_url,
               value           = excluded.value`,
        ).run(params);
    }
    return dense.length;
}

async function savePicks(
    rows: FcRow[], valueCol: string, rankCol: string, today: string,
): Promise<number> {
    let n = 0;
    for (const r of rows) {
        if (r.player.position !== 'PICK') continue;
        const m = PICK_LABEL.exec(r.player.name.trim());
        if (!m) continue;
        n += 1;
        // `query`, with an explicit RETURNING, and both halves of that
        // matter.
        //
        // getDb().run() appends `RETURNING id` to any INSERT that lacks one,
        // so it can report a lastInsertRowid. dynasty_picks is keyed on the
        // label and has no id column, so every pick insert died in
        // production on `column "id" does not exist` — while the player
        // writes beside them succeeded, because those go to `rankings`,
        // which does have an id. The cron reported 397 players saved and
        // quietly saved no picks at all.
        //
        // Naming a RETURNING stops the injection. It has to go through
        // `query` rather than `run` because better-sqlite3 refuses `.run()`
        // on a statement that returns rows, which is how the local path
        // would then have broken instead.
        await query(
            `INSERT INTO dynasty_picks
               (label, season, round, slot, ${valueCol}, ${rankCol}, scraped_at)
             VALUES ($1,$2,$3,$4,$5,$6,$7)
             ON CONFLICT (label) DO UPDATE SET
               season       = excluded.season,
               round        = excluded.round,
               slot         = excluded.slot,
               ${valueCol}  = excluded.${valueCol},
               ${rankCol}   = excluded.${rankCol},
               scraped_at   = excluded.scraped_at
             RETURNING label`,
            [r.player.name.trim(), Number(m[1]), Number(m[2]),
                m[3] ? m[3].toLowerCase() : null, r.value, r.overallRank, today]);
    }
    return n;
}

export interface DynastyRefreshResult {
    status: 'ok' | 'failed';
    scraped_at: string;
    [key: string]: unknown;
}

/** Refresh both formats. One failing must not cost the other its run. */
export async function refreshDynastyValues(today: string): Promise<DynastyRefreshResult> {
    await ensurePickTable();
    const pool = await query<{ id: number; sleeper_id: string | null }>(
        `SELECT id, sleeper_id FROM players
          WHERE sleeper_id IS NOT NULL AND sleeper_id != ''`);
    const bySleeper = new Map(pool.map(p => [String(p.sleeper_id), p.id]));

    const settled = await Promise.allSettled(DYNASTY_FEEDS.map(async feed => {
        const rows = await fetchFeed(feed.query);
        const players = await savePlayers(rows, feed.source, bySleeper, today);
        const picks = await savePicks(rows, feed.valueCol, feed.rankCol, today);
        return { key: feed.key, source: feed.source, players, picks };
    }));

    const report: Record<string, unknown> = {};
    settled.forEach((r, i) => {
        const feed = DYNASTY_FEEDS[i];
        report[feed.key] = r.status === 'fulfilled'
            ? { source: r.value.source, players: r.value.players, picks: r.value.picks }
            : { error: String((r.reason as Error)?.message ?? r.reason) };
    });

    return {
        status: settled.some(r => r.status === 'fulfilled') ? 'ok' : 'failed',
        scraped_at: today,
        matchable: bySleeper.size,
        ...report,
    };
}

File: dynasty-scout/scrapers/rankings/dynasty_values.py
"""
scrapers/rankings/dynasty_values.py

Dynasty trade values for every player, and for draft picks.

Why this exists alongside fantasycalc.py, which hits the same API: that one
asks for `rookiesOnly=true` and then drops everything whose position is
"PICK". It was written for a rookie scouting board, where both of those are
the right call. A dynasty trade calculator needs precisely the two things it
throws away — Ja'Marr Chase has to be priceable, and so does a 2027 1st.

It writes to its own sources rather than reusing "FantasyCalc". That source
is rookie ranks 1..N and run_consensus.py weights it at 0.30 of the rookie
consensus; overwriting it with overall dynasty ranks would quietly corrupt
the rookie board this app was built to be. The rookie sources are a snapshot
of how the 2026 class was seen before the draft, and they are meant to stay
that way — nothing here writes to them.

Players are matched on sleeperId, not on names. Every one of the 397 players
the API returns matches that way, so the fuzzy-name path the rookie scraper
needs is not needed here.

Run: py scrapers/rankings/dynasty_values.py
"""

import os
import re
import sqlite3
import time
from datetime import date

import requests

DB_PATH = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 'dynasty_scout.db')

BASE = "https://api.fantasycalc.com/values/current"
FEEDS = [
    # (source name, query, which value column the picks table gets)
    ("FantasyCalc Dynasty",
     "?isDynasty=true&numQbs=1&ppr=1&superflex=false", "value_1qb", "rank_1qb"),
    ("FantasyCalc Dynasty SF",
     "?isDynasty=true&numQbs=2&ppr=1&superflex=true", "value_sf", "rank_sf"),
]
SOURCE_URL = "https://fantasycalc.com/rankings"

# "2027 1st (Early)", "2028 2nd", ...
PICK_LABEL = re.compile(
    r"^(\d{4}) (\d+)(?:st|nd|rd|th)(?: \((Early|Mid|Late)\))?$")

PICKS_DDL = """
CREATE TABLE IF NOT EXISTS dynasty_picks (
  id         INTEGER PRIMARY KEY AUTOINCREMENT,
  label      TEXT NOT NULL UNIQUE,   -- '2027 1st (Early)'
  season     INTEGER NOT NULL,
  round      INTEGER NOT NULL,
  -- 'early' | 'mid' | 'late', or NULL for the unspecified pick, which is
  -- what you hold before the standings say where it lands.
  slot       TEXT,
  value_1qb  INTEGER,
  value_sf   INTEGER,
  rank_1qb   INTEGER,
  rank_sf    INTEGER,
  scraped_at TEXT NOT NULL
)
"""


def fetch(query: str):
    r = requests.get(BASE + query, timeout=30)
    r.raise_for_status()
    return r.json()


def save_players(cur, rows, source, player_by_sleeper, today):
    """Dynasty values for everybody the API knows and we also know."""
    matched, unmatched = 0, []
    for item in rows:
        pl = item["player"]
        if pl.get("position") == "PICK":
            continue
        pid = player_by_sleeper.get(str(pl.get("sleeperId") or ""))
        if not pid:
            unmatched.append(pl["name"])
            continue
        matched += 1
        # Appended by date, not replaced. The redraft sources already keep a
        # row per scrape — five dates each — and a dynasty value is worth the
        # same treatment: what a player was worth in September is how you
        # tell whether a trade you are being offered has moved. Re-running on
        # the same day overwrites that day and nothing else, which is what
        # the unique key is for.
        cur.execute(
            """INSERT INTO rankings
               (player_id, source, rank_overall, rank_positional, value,
                source_url, scraped_at)
               VALUES (?,?,?,?,?,?,?)
               ON CONFLICT(player_id, source, scraped_at) DO UPDATE SET
                 rank_overall=excluded.rank_overall,
                 rank_positional=excluded.rank_positional,
                 value=excluded.value,
                 source_url=excluded.source_url""",
            (pid, source, item.get("overallRank"), item.get("positionRank"),
             item.get("value"), SOURCE_URL, today))
    print(f"  {source}: {matched} players"
          + (f", {len(unmatched)} unmatched {unmatched[:5]}" if unmatched else ""))
    return matched


def save_picks(cur, rows, value_col, rank_col, today):
    """
    Draft picks, which are assets in a dynasty trade and not players.

    Upserted by label so a pick keeps its id across runs — the calculator
    refers to them by id, and a link somebody sent last week should still
    open the same pick this week.
    """
    seen = 0
    for item in rows:
        pl = item["player"]
        if pl.get("position") != "PICK":
            continue
        m = PICK_LABEL.match(pl["name"].strip())
        if not m:
            print(f"  ! unparsed pick label: {pl['name']!r}")
            continue
        season, rnd, slot = m.group(1), m.group(2), m.group(3)
        seen += 1
        cur.execute(
            f"""INSERT INTO dynasty_picks
                  (label, season, round, slot, {value_col}, {rank_col}, scraped_at)
                VALUES (?,?,?,?,?,?,?)
                ON CONFLICT(label) DO UPDATE SET
                  season=excluded.season,
                  round=excluded.round,
                  slot=excluded.slot,
                  {value_col}=excluded.{value_col},
                  {rank_col}=excluded.{rank_col},
                  scraped_at=excluded.scraped_at""",
            (pl["name"].strip(), int(season), int(rnd),
             slot.lower() if slot else None,
             item.get("value"), item.get("overallRank"), today))
    print(f"  picks: {seen} ({value_col})")
    return seen


def run():
    conn = sqlite3.connect(DB_PATH, timeout=15)
    cur = conn.cursor()
    cur.execute(PICKS_DDL)

    cur.execute("SELECT sleeper_id, id FROM players "
                "WHERE sleeper_id IS NOT NULL AND sleeper_id != ''")
    player_by_sleeper = {str(s): i for s, i in cur.fetchall()}
    today = date.today().isoformat()

    print("Fetching FantasyCalc dynasty values (all players + picks)...")
    for i, (source, query, value_col, rank_col) in enumerate(FEEDS):
        rows = fetch(query)
        save_players(cur, rows, source, player_by_sleeper, today)
        save_picks(cur, rows, value_col, rank_col, today)
        conn.commit()
        if i + 1 < len(FEEDS):
            time.sleep(1)

    n_players = cur.execute(
        "SELECT COUNT(DISTINCT player_id) FROM rankings "
        "WHERE source LIKE 'FantasyCalc Dynasty%'").fetchone()[0]
    n_picks = cur.execute("SELECT COUNT(*) FROM dynasty_picks").fetchone()[0]
    print(f"done: {n_players} players and {n_picks} picks carry a dynasty value")
    # Fold the write-ahead log back into the .db before closing. The app
    # opens SQLite in WAL mode; the -wal file is gitignored and the .db is
    # tracked, so without this a run can commit its work and still ship a
    # database file that does not contain it.
    try:
        conn.execute("PRAGMA wal_checkpoint(TRUNCATE)")
    except Exception as e:
        print(f"  ! could not checkpoint the WAL: {e}")
    conn.close()


if __name__ == "__main__":
    run()
